opentelemetry-resource-detectors/src/host.rs

```diff
@@ -0,0 +1,71 @@
+//! HOST resource detector
+//!
+//! Detect the unique host ID.
+use std::error::Error;
+use std::fs::read_to_string;
+use std::path::Path;
+use std::time::Duration;
+use opentelemetry::KeyValue;
+use opentelemetry_sdk::Resource;
+use opentelemetry_sdk::resource::ResourceDetector;
+
+/// Detect the unique host ID.
+///
+/// This detector looks up the host id using the sources defined
+/// in the OpenTelemetry semantic conventions [`host.id from non-containerized systems`].
+///
+/// [`host.id from non-containerized systems`]: https://opentelemetry.io/docs/specs/semconv/resource/host/#collecting-hostid-from-non-containerized-systems
+pub struct HostResourceDetector {
+    host_id_detect: fn() -> Option<String>
+}
+
+impl ResourceDetector for HostResourceDetector {
+    fn detect(&self, _timeout: Duration) -> Resource {
+        (self.host_id_detect)().map(|host_id| {
+            Resource::new(vec![KeyValue::new(
+                opentelemetry_semantic_conventions::resource::HOST_ID,
+                host_id,
+            )])
+        }).unwrap_or(Resource::new(vec![]))
+    }
+}
+
+#[cfg(target_os = "linux")]
+fn host_id_detect() -> Option<String> {
+    let machine_id_path  = Path::new("/etc/machine-id");
+    let dbus_machine_id_path = Path::new("/var/lib/dbus/machine-id");
+    read_to_string(machine_id_path).or_else(|_|{read_to_string(dbus_machine_id_path)}).ok()
+}
+
+#[cfg(not(target_os = "linux"))]
```

Comment:
Should we addd a TODO here for support of non-linux platforms?

Reply:
Fixed in last commit https://github.com/open-telemetry/opentelemetry-rust-contrib/pull/57/commits/1b0fa9174abb504a7028d3f39ca501b468331022